Followers section in People Management

oguzkocer opened the pull request:
Closes #4142.

**Decisions**
* I decided to disable swipe to refresh for this feature. Syncing is always an issue for these type of features and to avoid some issues, we implemented a way to remove all the people except for the first page when you first land on the people list. Enabling refresh would make things more complicated and would make our previous decision meaningless. Also, people list should be fairly static and when you go back to My Site and come back to people list, it'll be refreshed.

**Known issues:**
* I recently found some RTL issues while testing the feature: https://cloudup.com/cObeMJEy0yv I'll be looking into this issue, but thought it shouldn't hold the code review.
* There is a possible design change for the subscriber date position, currently waiting feedback from @rickybanister
* While testing a large list of followers, I noticed that when you go into detail and come back you lose your scroll position. I've tried many times to reproduce this, but was unable to do so. I'll keep an eye on this one and please let me know if you notice anything similar.

**Questions**
* If there is an error, we currently show a toast as well as set the empty view message to that error. I am not sure if the duplicate approach is good here. Should we remove that toast message?
* I feel like the loading progress bar, for infinite scroll, should be centered. However, that component is used on other pages as well, so we should make a decision on whether we should move it for all those pages or keep it as is.

**To test:**
* Go into the People section from My Site page and check your followers & email followers
* It'd be great if we install a version from `develop` first and then update it from this branch to test the migration. The migration is pretty minimal, but it's a good check. Make sure you visit the people list when you first install it from `develop` so the db will have some data to work with.
* There is a lot of state management going on, so it'd be good to test network availability, empty lists for followers & email followers, loading the list for the first time, loading more followers etc.
* If you have a site with many followers, it's a good test as well to make sure our infinite scroll is working correctly.
* Go into the detail page of a follower and try to remove her.
* While trying to add an email follower to your site for test purposes, keep in mind that the email can not belong to a WordPress.com user. You can always use the `youremail+something@example.com` gmail trick for this.

/cc @hypest, @astralbodies 

@mzorz as we discussed on Slack a while ago, I am pinging you for review since you built the original version of `FilteredRecyclerView` and most of this PR is about that. Do let me know if you have any questions about the specifics of People Management and if there is anything not clear at first glance so I can add comments accordingly. Thanks!

Needs review: @mzorz

oguzkocer commented:
> I recently found some RTL issues while testing the feature: https://cloudup.com/cObeMJEy0yv I'll be looking into this issue, but thought it shouldn't hold the code review.

I've found out that the issue was related to how I setup RelativeLayout, https://github.com/wordpress-mobile/WordPress-Android/pull/4313/commits/01c07fefe1da99693890332b5fa35d5c31548064 should fix that.

oguzkocer commented:
I've started working on the `Viewers` section in `feature/4312-viewer-list` branch and I quickly realized that we would need another migration for the `is_viewer` column. Since the review for this PR hasn't started yet, I decided to add the relevant commit https://github.com/wordpress-mobile/WordPress-Android/pull/4313/commits/07744a8e6479040cd94543c0bbf8eccd22973671 in this PR instead. This change also introduces one more column to primary key which changes a lot of the SQL queries and I think it's better to test them all together.

mzorz commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleListFragment.java:
```diff
@@ -71,60 +78,173 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
+
+            @Override
+            public String onShowEmptyViewMessage(EmptyViewMessageType emptyViewMsgType) {
+                if (emptyViewMsgType == EmptyViewMessageType.NO_CONTENT) {
```
minor thing here: you could have this case implemented inside the switch below, instead of having to make a distinction between `EmptyViewMessageType.NO_CONTENT` and the rest of the cases. This case should be handled by ` switch (emptyViewMsgType) `

mzorz commented on WordPress/src/main/java/org/wordpress/android/ui/people/PersonDetailFragment.java:
```diff
@@ -105,11 +118,31 @@ public void refreshPersonDetails() {
             mRoleTextView.setText(StringUtils.capitalize(person.getRole()));
 
-            setupRoleContainerForCapability();
+            if (!person.getUsername().isEmpty()) {
```
It would be better to use `TextUtils.isEmpty(person.getUsername())` to prevent a possible NPE from happening

mzorz commented:
Ok so here’s my feedback @oguzkocer , please let me know if you find that I seem to be misunderstanding something so I can re-review :)

## General

- I’ve tested `develop` first, then installed your branch, and confirmed the database upgrade is working fine :+1:

- I’ve seen the list of followers don’t have any actions on long press, but then if I tap on each follower (except for myself) then I can remove them. I think the UX would benefit from having a way to delete right away, and also having a way to mark several items and delete them all, like the way it’s done in the Comments section of the app.
(see screenshot)
<img width="611" alt="screen shot 2016-06-30 at 2 34 07 pm" src="https://cloud.githubusercontent.com/assets/6597771/16503800/22f68e3c-3eeb-11e6-8005-6da219571759.png">

- When trying to add a follower, the “Role” area would benefit from a down arrow that indicated this is a touchable material to represent an actionable item

- The Calypso web admin interface also has a link right there that takes you to a page with more information about roles, called “Learn more about roles”. It would be nice to have such a link there. (see screenshot)
<img width="697" alt="screen shot 2016-06-30 at 2 37 13 pm" src="https://cloud.githubusercontent.com/assets/6597771/16503817/3324fd0c-3eeb-11e6-86a5-5d745532fca6.png">

- When entering a username and hitting SEND; there’s a strange cursor movement upon validation. I wonder what that is, but that doesn’t seem proper handling
![people_strange_validation](https://cloud.githubusercontent.com/assets/6597771/16504303/b6e1e4c8-3eed-11e6-8471-bbed5871a2e7.gif)

- When entering a valid username and hitting SEND and connection is available, it works quickly and a toast is shown (Invitation Sent or alike). However, when switched my internet connection off, then entered a valid username (nbradbury) then hit SEND, and nothing happens, so I can’t know if the app even tried sending, if it is still trying, or what. I think the “SEND”  option should be replaced by a spinning wheel to indicate the app is trying to process the invitation, and as soon as your call returns (either with error or successfully) then hide then spinner wheel and show the SEND option again, and show a toast with the result / inviting the user to try again if it was unsuccessful.

- I like the detail that the Custom Message count will read “1 character remaining” (singular) as opposed to “0/n characters remaining” (plural) when count != 1. :)


## Design considerations (please run these through a designer too):

- Design wise, the fact that we have a continuation of the same surface from toolbar to item view in the user detail view (see screenshot) seems odd to me. It seems to be there on purpose as per the comment on this line https://github.com/wordpress-mobile/WordPress-Android/blob/feature/4142-people-list-dropdown/WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java#L370
From my point of view, a toolbar separator should be shown in there, but might be wise to bring in the eye of a designer here.

- In terms of the workflow, the PersonDetailFragment (detail view) doesn’t seem to be showing that much. Also, the only doable action an item has (which is, remove a follower)  can only be accessed through the Detail view, which doesn’t really offer anything else that what is shown in the list itself. I wonder if we can do without it and have the trash bin incorporated to each item in the list, on the right side.


## Code

I made a few comments here and there in the code, and also wanted to note this:

- Good call on having the listeners set with `setOnPersonSelectedListener` both when the activity is created and again `onCreate(saved instance)` as having the activity be a listener of something that happens within the Fragment is always a painpoint
(https://github.com/wordpress-mobile/WordPress-Android/blob/feature/4142-people-list-dropdown/WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java#L132)

- I wonder swipe-to-refresh has been taken out? If you load the data from the server and then save to the DB, then only show on the screen whatever comes from the DB, my guess would be you’ll always have a consistent state. Was there any other problem we could deal with?

oguzkocer commented:
Thanks for the feedback @mzorz!

**Pull to refresh**

I don't think it's as simple as that. You still need to consider the cases like keyboard rotations, starting from background etc and in all of those cases you'll get `onLoadData` and just to avoid refreshing the list more than once, I introduced variables like `mCanRefreshUsers`. We still can implement it with some refactoring to the `FilteredRecyclerView` to force a refresh when it's forced by the user, but I don't think the benefit outweighs the possible issues. The biggest reason I am opposed to this is that, if the user scrolls even a little bit and a refresh is completed, there'll be a jump and there is no way to get around that without introducing syncing issues. That alone is good enough a reason not to introduce this feature especially for a static list. Hope this clears the reasoning behind that decision. Do let me know if I am mistaken on any of these points.

**DB refactor**

As discussed on Slack, I've separated the `PeopleTable` into 3 different tables (one for each filter). This change greatly simplified the queries we used and will be much easier to maintain. I've decided to keep the `PeopleTable` class and hide the multiple table implementation from the users of that class. You'll need to **uninstall** your current version, since I changed the current migration instead of adding a new one. This is a really big change and it touched a lot of the codebase, so I am hoping I didn't introduce any bugs. It'd be great if you can watch out for it as well, thanks!

**Person model refactor**

While I was updating the DB layer for the people management feature as a whole, I decided to drop some unused properties such as `firstName` & `lastName`. I've also dropped the remote blog id property. The problem with that property was not all endpoints were returning that value and it's meaning is not always clear. Let's say you have a follower and it has a blog id, is that your blog's id or is it follower's own blog id? So, to avoid developer errors, I thought it's best to just drop that property. Local blog id always refer to which blog that Person belongs to and you can easily access the information either using the local id or just getting the current blog id which is always available. I think this change makes things more clear and less error-prone.

---

These are all the changes I've worked on since Friday, I've done my best to review & test the changes. I am really sorry for such a big change after the PR was submitted, but I really think the changes were for the best and should make the second pass easier. Thanks again for your patience and feedback, it's greatly appreciated!

mzorz commented on WordPress/src/main/java/org/wordpress/android/datasets/PeopleTable.java:
```diff
@@ -48,122 +71,214 @@ public static void reset(SQLiteDatabase db) {
-            getWritableDb().delete(PEOPLE_TABLE, "local_blog_id=?1 AND person_id " +
-                    "IN (SELECT person_id FROM " + PEOPLE_TABLE + " WHERE local_blog_id=?1 " +
-                    "ORDER BY display_name DESC LIMIT ?2)", args);
+        getWritableDb().delete(table, "local_blog_id=?1", args);
```
TIL - didn't know about the `?1` notation, but guess it's not necessary here with only one argument. No need to take action though. 👍

mzorz commented:
Ok on the pull to refresh decision @oguzkocer 👍 
Have been looking here and there and thought I found a couple of issues but in the end my doubts were cleared by code itself. Looks good!

mzorz commented:
:shipit:
